lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,662 @@
+
+    /**
+     * Configuration store to use.
+     *
+     * @var TransformerConfiguration
+     */
+    private $configuration;
+
+    /**
+     * Reference node to attach preload links to.
+     *
+     * @var DOMElement|null
+     */
+    private $preloadReferenceNode;
+
+    /**
+     * Inline style backup attribute that stores inline styles that are being moved to <style amp-custom>.
+     *
+     * An empty string signifies that no inline style backup is available.
+     *
+     * @var string
+     */
+    private $inlineStyleBackupAttribute;
+
+    /**
+     * Instantiate a PreloadHeroImage object.
+     *
+     * @param TransformerConfiguration $configuration Configuration store to use.
+     */
+    public function __construct(TransformerConfiguration $configuration)
+    {
+        $this->configuration = $configuration;
+    }
+
+    /**
+     * Apply transformations to the provided DOM document.
+     *
+     * @param Document        $document DOM document to apply the transformations to.
+     * @param ErrorCollection $errors   Collection of errors that are collected during transformation.
+     * @return void
+     */
+    public function transform(Document $document, ErrorCollection $errors)
+    {
+        if ($this->configuration->get(PreloadHeroImageConfiguration::PRELOAD_HERO_IMAGE) === false) {
+            return;
+        }
+
+        $this->inlineStyleBackupAttribute = $this->configuration->get(
+            PreloadHeroImageConfiguration::INLINE_STYLE_BACKUP_ATTRIBUTE
+        );
+
+        $heroImages     = $this->findHeroImages($document);
+        $heroImageCount = count($heroImages);
+        if ($heroImageCount > self::DATA_HERO_MAX) {
+            $errors->add(Error\TooManyHeroImages::whenPastMaximum());
+            $heroImageCount = self::DATA_HERO_MAX;
+        }
+
+        for ($index = 0; $index < $heroImageCount; $index++) {
+            $this->generatePreload($heroImages[$index], $document, $errors);
+            $this->generateImg($heroImages[$index], $document);
+        }
+    }
+
+    /**
+     * Find the hero images to optimize.
+     *
+     * @param Document $document Document to look for hero images in.
+     * @return HeroImage[] Array of hero images to optimize.
+     */
+    private function findHeroImages(Document $document)
+    {
+        $heroImageCandidate = null;
+        $heroImages         = [];
+        $node               = $document->body;
+
+        while ($node !== null) {
+            if (! $node instanceof DOMElement) {
+                $node = $this->nextNode($node);
+                continue;
+            }
+
+            $heroImage = $this->detectImageWithDataHero($node);
+            if ($heroImage) {
+                $heroImages[] = $heroImage;
+            }
+
+            if (! $heroImageCandidate && count($heroImages) === 0) {
+                $heroImageCandidate = $this->detectHeroImageCandidate($node);
```

Comment:
Together with the `DetermineHeroImage` transformer, we will actually have a `data-hero` attribute for WordPress in almost all scenarios.

Reply:
Also, if we don't, it means that there's no site logo and no featured image available. I don't think we should use two images in that case, as it will very probably be main content only.

Reply:
Oh got it. So `DetermineHeroImage` will have already run. 👍